### PR TITLE
Parser v2 example

### PR DESCRIPTION
**Pull request by aljazkonec1**
This PR adds an updated Parser v2 example. New additions are: 

- A base `Parser` class that all other classes will inherit. 
- Updated sytax using `.build()` to be more in line with dai v3 syntax, the .build method accepts a list of dai.nn_archive.v1.Head objects or a Dictionary. If Head objects are parsed, they are converted to a dictionary.

Docstrings will be updated once the functionality is finalized. Variable names also can be adjusted.

### CODE REVIEW

**Comment by kkeroo** [comment]
I am thinking if we can also move parser-specific parameter setting to build method. E.g. ClassificationParser requires `classes` parameter. Instead of reading it in the parser we can do it inside `build` from BaseParser.

```
class BaseParser:
    def build(self, params: dict):
        # Loop through the dictionary and set attributes
        for key, value in params.items():
            if hasattr(self, key):
                setattr(self, key, value)
            else:
                print(f"Warning: {key} is not a valid attribute")

class ClassificationParser(BaseParser):
    def __init__(self):
        self.param1 = None
        self.param2 = None
        # You can define more default parameters here if needed

# Example usage:
params_dict = {'param1': 'value1', 'param2': 'value2'}

parser = ClassificationParser()
parser.build(params_dict)

print(parser.param1)  # Output: value1
print(parser.param2)  # Output: value2
```

This is quick GPT generated code, but I think it is possible. So, we map all parameters from head->metadata to parser's parameters. If there is no parser parameter we print a warning.

**Comment by aljazkonec1** [comment]
I'm not sure this is going to work because in dai it wil get ported to C, which doesnt support setattr()

File: depthai_nodes/ml/parsers/parser.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        self.head_configs: Dict = {}
+
+    def build(self, head_metadata: Union[List, Dict], name: str = ""):
+        """Initial build method for all parsers."""
+
+        if isinstance(head_metadata, list):
+            if len(head_metadata) == 0:
+                raise ValueError("No heads parsed from archive")
+            if len(head_metadata) > 1 and name == "":
+                raise ValueError("Multiple heads detected, please specify head name")
```
Comment (klemen1999) [inline depthai_nodes/ml/parsers/parser.py:23]: Can we make this even smarter: 
If we are initializing ClassificationParser and only one head has `ClassificationParser˙ set as `parser` then we can automatically pick that one even if we get a whole list of all heads passed in.
Something like this:
```
current_parser = self.__class__.__name__ # this should return "ClassificationParser"
all_heads_parsers = [h.parser for h in head_metadata]
num_matches = all_heads_parsers.cound(current_parser)
if num_matches == 1:
    # good, take that one
else:
    # need head name
```

**Review by klemen1999 — COMMENTED** [review]
The structure LGTM, left one suggestion

File: depthai_nodes/ml/parsers/parser.py (hunk 1 above)
Comment (aljazkonec1) [reply depthai_nodes/ml/parsers/parser.py:23]: Sounds good, I'll add it.

File: depthai_nodes/ml/parsers/classification.py [hunk 2]
```diff
@@ -71,16 +108,7 @@ def run(self):
...
-            if len(output_layer_names) != 1:
-                raise ValueError(
-                    f"Expected 1 output layer, got {len(output_layer_names)}."
-                )
-
-            if self.n_classes == 0:
-                raise ValueError("Classes must be provided for classification.")
-
-            scores = output.getTensor(output_layer_names[0], dequantize=True).astype(
+            scores = output.getTensor(self.output_layer_name, dequantize=True).astype(
```
Comment (jkbmrz) [inline depthai_nodes/ml/parsers/classification.py:111]: `self.output_layer_name` is only defined by calling the `build()` method. We are missing a way to define it manually.

**Review by jkbmrz — COMMENTED** [review]
Adding some more suggestions. Moreover, we are missing docstrings / explanation of parameters. It’s good to add this now so that the reviewers are 100% sure what they look at (cc @aljazkonec1). I'm getting a bit lost in the current parameter naming in the `Parser` class and would rename as follows:
- `head_configs`: Is this a “config” of a single head? If yes, shouldn’t it be named `head_config`? Additionally, not really sure if this is the best name for it.
- `head_metadata`: If I understand correctly this is it metadata of one head or multiple heads (if existent). So I’d rather name it as `heads_metadata` or just `heads` (we iterate through it later in the code and call each iteration a head). Also, if `head_metadata` is a list of single dict, how is this different from the `head_configs`?
- `name`: I’d rename it to `head_name` for better readability.

File: depthai_nodes/ml/parsers/classification.py [hunk 3]
```diff
@@ -31,21 +32,57 @@ class ClassificationParser(dai.node.ThreadedHostNode):
...
+            print(
+                "No classes provided in nn_archive metadata. Please provide class names in the nn_archive"
+            )
+
+        try:
+            self.n_classes = self.head_configs["n_classes"]
+        except KeyError:
+            print(
+                "No n_classes provided in nn_archive metadata. Please provide number of classes in the nn_archive"
+            )
```
Comment (aljazkonec1) [inline depthai_nodes/ml/parsers/classification.py:82]: Are these OK KeyErrors? Or should they be adjusted?

File: depthai_nodes/ml/parsers/parser.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                num_matches = parser_names_in_archive.count(current_parser)
+                if num_matches == 0:
+                    raise ValueError(
+                        f"No heads available for {current_parser} in the nn_archive."
+                    )
+                elif num_matches == 1:
+                    head = [head for head in heads if head.parser == current_parser][0]
+                else:
+                    raise ValueError(
+                        "Multiple heads detected, please specify a head name."
```
Comment (klemen1999) [inline depthai_nodes/ml/parsers/parser.py:52]:
```suggestion
                        f"Multiple heads with parser={current_parser} detected, please specify a head name."
```

**Review by klemen1999 — APPROVED** [review]
LGTM

File: depthai_nodes/ml/parsers/classification.py [hunk 5]
```diff
@@ -20,32 +21,70 @@ class ClassificationParser(dai.node.ThreadedHostNode):
...
+        self.output_layer_name: str = ""
+        self.classes: List = None
+        self.n_classes: int = 0
+        self.is_softmax: bool = True
+
+    def build(
+        self,
+        heads: Union[List, Dict],
+        head_name: str = "",
+        is_softmax: bool = True,
```
Comment (kkeroo) [inline depthai_nodes/ml/parsers/classification.py:61]: This also should be in nn archive like `classes` and `n_classes`

**Review by kkeroo — APPROVED** [review]

File: depthai_nodes/ml/parsers/classification.py (hunk 3 above)
Comment (kkeroo) [reply depthai_nodes/ml/parsers/classification.py]: Hm, This does not error out, right? But just prints out the warning. So, `setClasses` method should be used?

File: depthai_nodes/ml/parsers/parser.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+            head_dictionary = {}
+            head_dictionary["parser"] = parser_name
+            head_dictionary["outputs"] = outputs
+            if metadata is not None:
+                head_dictionary.update(metadata.extraParams)
+            self.head_config = head_dictionary
+
+        else:
+            self.head_config = heads
```
Comment (jkbmrz) [inline depthai_nodes/ml/parsers/parser.py:88]: I'd rather write this as `elif isinstance(heads, dict)` as type hints do not enforce type checking at runtime. So it's possible one provides a heads an int and this code won't return an error. Thus, I'd also add an `else` statement that raises a `ValueError` that the provided heads object is neither a list or a dict.

File: depthai_nodes/ml/parsers/parser.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+class Parser(dai.node.ThreadedHostNode):
+    """Base class for all parsers."""
+
+    def __init__(self):
+        super().__init__()
+        self.input = self.createInput()
+        self.out = self.createOutput()
+
+        self.head_config: Dict = {}
```
Comment (jkbmrz) [inline depthai_nodes/ml/parsers/parser.py:14]: We miss a docstring explanation of the `head_config` parameter.

**Review by jkbmrz — COMMENTED** [review]
Adding 2 comments to fix in the next PR.

File: depthai_nodes/ml/parsers/classification.py [hunk 8]
```diff
@@ -20,32 +21,57 @@ class ClassificationParser(dai.node.ThreadedHostNode):
...
 
-    def __init__(self, classes: List[str] = None, is_softmax: bool = True):
+    def __init__(self):
         """Initializes the ClassificationParser node.
 
-        @param classes: List of class names to be used for linking with their respective
-            scores.
-        @param is_softmax: If False, the scores are converted to probabilities using
-            softmax function.
+        Attributes
```
Comment (tersekmatija) [inline depthai_nodes/ml/parsers/classification.py:38]: It makes sense to document that those only get populated after `build()` is called.

**Pull request merged**